[class-inheritance] Concept Documentation (#2960)
exercism/python: concepts/class-inheritance/about.md

# About

Inheritance is one of the ['four pillars'][four-pillars] of Object Oriented Programming (`OOP`).
In situations where only a small amount of functionality needs to be customized for a new class, `inheritance` allows code re-use from one or more parent classes, and can help make programs cleaner and more maintainable.

## Inheritance

> `Inheritance` describes `is a kind of` relationship between two or more classes, abstracting common details into super class and storing specific ones in the subclass.

Reviewer: ```suggestion
`Inheritance` describes `is a kind of` relationship between two or more classes, abstracting common details into super (or _parent_) class and storing specific ones in the subclass (or _child class_).

```

To create a child class, specify the parent class name inside the pair of parenthesis, followed by it's name.
Example
```python
class Child(Parent):  
   pass
```

> Every child class inherits all the behaviours exhibited by their parent class.

Reviewer: ```suggestion
Every child class inherits all the behaviors (_attributes, constructors, methods_) exhibited by their parent class.
 
```

## Single Inheritance

> When a derived (or child) class inherits only from one base (or parent) class, it is known as single inheritance.

Reviewer: ```suggestion
When a derived (or child) class inherits only from one base (or parent) class, it is known as _single inheritance_.

```

```python

> class Person:

Reviewer: ```suggestion
# The parent or base class.
class Person:

```

    def __init__(self, fname, lname):
        self.fname = fname

>         self.lname = lname
> class Employee(Person):

Reviewer: ```suggestion
        self.lname = lname
        
class Employee(Person):

```

    all_employees = []
    def __init__(self, fname, lname, empid):

>         Person.__init__(self, fname, lname)

Reviewer: ```suggestion
        # Using the Parent constructor to create the base object.
        Person.__init__(self, fname, lname)

```

>         self.empid = empid

Reviewer: ```suggestion
        # Adding an attribute specific to the Child class.
        self.empid = empid

```

        Employee.all_employees.append(self)
```
`Employee` class is derived from `Person`.

> Now, we can create an `Employee` object.

Reviewer: ```suggestion
Now, we can create an `Employee` object.

```

```python
...
p1 = Person('George', 'smith')
print(p1.fname, '-', p1.lname)
e1 = Employee('Jack', 'simmons', 456342)
e2 = Employee('John', 'williams', 123656)
print(e1.fname, '-', e1.empid)
print(e2.fname, '-', e2.empid)
```
After running the program we will get the following output

> ```cmd

Reviewer: ```suggestion
```bash

```

George - smith
Jack - 456342
John - 123656
```
## Multiple Inheritance

> As we've seen, a class inherits from another class in this scenario. On the other side, multiple inheritance is a feature that allows a class to inherit characteristics and methods from many parent classes.

Reviewer: ```suggestion
As we've seen, `single inheritance` is where a class inherits directly from another class. 
On the other side, `multiple inheritance` is a Python feature that allows a child class to inherit characteristics and methods from more than one parent class.

```

```python
class SubclassName(BaseClass1, BaseClass2, ...):
    pass
```

> ### Diamond Problem

Reviewer: ```suggestion
### Multiple Inheritance and the Diamond Problem

```

# About

> A `Class` is a template/blueprint representing a group of objects that share common properties and relationships.

Author: Ok

         print("Roll: ", self.__roll)
         print("Marks:", self.__marks)
         print("Division: ", self.__div)
         
     # Accessor Function    
     def getRollNo(self):
        return self.__roll
     
     # Mutator Function
     def setDiv(self):
        if self.__marks>=60:
            self.__div = 'A'
        elif self.__marks<60:
            self.__div = 'B'
        return self.__div
```
__2. Mutator Function__: This method is used to change the state of an object, i.e. it changes the data variable's hidden value. It has the ability to change the value of a variable on the fly. The update method is another name for this function. We can also use the word _set_ to name these procedures.

As per the above example, `getRollNo()` is an accessor function and `setDiv` is a mutator function.

## Inheritance

## Single Inheritance

When a subclass inherits only from one base class, it is known as single inheritance.

```python
#Base class
class Odin:
    def func1(self):
        print("Odin: father of thor")

#Derived/Subclass class
class Thor(Odin):
    def func2(self):
        print("Thor: son of Odin")

object = Thor()
object.func1()
object.func2()
```

> The derived class will print both the statements of `func1` and `func2`.

Reviewer: I think this needs to be a more direct and functional example that uses both attributes and methods, to demonstrate _what_ gets inherited, and _how_ one then adds to or overrides those properties and functionality in the child class.  Recommend taking a look at [this DO article](https://www.digitalocean.com/community/tutorials/understanding-class-inheritance-in-python-3) for some inspiration and examples.   This [real python](https://realpython.com/python3-object-oriented-programming/#inherit-from-other-classes-in-python) article is good as well.

Author: You are right, I was also getting a feeling that the example wasn't upto the mark. I have written a new example.

## Multiple Inheritance
As we've seen, a class inherits from another class in this scenario. On the other side, multiple inheritance is a feature that allows a class to inherit characteristics and methods from many parent classes.
```python
class SubclassName(BaseClass1, BaseClass2, ...):
    pass

> ```

Reviewer: I think this section needs to be extended to talk about how multiple inheritance works -- and how the parent classes are read in.  I think it might also be helpful to limit the initial example to only two base classes - one the "core" class, and one the "secondary", to help lead into the diamond problem and then to why mixins are a thing.

As with the single inheritance section above, I think it's important here to show how class attributes, instance attributes, and methods are inherited, and from there, how those get extended or overridden in the child class, so more code examples, and more narrative.

### Diamond Problem
The "diamond problem" (also known as the "deadly diamond of death") refers to an ambiguity that occurs when two classes B and C inherit from a superclass A, while another class D inherits from both B and C. If A has a method "m" that B or C (or even both of them) has overridden, and if it does not override this method, the question becomes which version of the method D inherits. It's possible that it's from A, B, or C.
Let's have a look at the problem using an example:

```python
class A:
    def m(self):
        print("m of A called")
class B(A):
    def m(self):
        print("m of B called")
class C(A):
    def m(self):
        print("m of C called")
class D(B,C):
    pass
```
If we call an instance x of class D, we will get the output as `m of B called`. But if we interchange the order of inheritance in class D i.e. `Class D(C, D)`. We will get the output as `m of C called`. 

> To solve the diamond problem in python, we will look into a new method `mro()`.

Reviewer: This isn't a new method, this is how Python solves the diamond problem.  `mro()` exposes what Python is doing, but it doesn't turn it on or off.  Suggest wording closer to what is [here](https://www.educative.io/edpresso/what-is-mro-in-python)

For example 
```python
class A1(object):
    def method(self):
        return 1

class A2(object):
    def method(self):
        return 2

class B1(object):
    def usesMethod(self):
        return self.method() + 10

class B2(object):
    def usesMethod(self):
        return self.method() + 20

class C1_10(A1, B1): pass
class C1_20(A1, B2): pass
class C2_10(A2, B1): pass
class C2_20(A2, B2): pass
```
Mixins helps us to recombine functionalities with different choices of base classes.
#### Pros and Cons of Mixins
| Advantages | Disadvantages |
|:-- | :-- |
|Mixin classes tend to be simple because they represent simple orthogonal concepts. | Execution of statements at run time tends to jump around in different mixins, making it hard to follow and debug|
|Helps us to recombine functionalities with different choices | Potential for long compile times|
## __super()__
In a nutshell, `super()` gives us access to methods in a superclass from the subclass that inherits from it.
`super()` by itself returns a temporary object of the superclass, which may subsequently be used to call the methods of that superclass.

But why we want to use `super()`?

> Using `super()` to call already created methods avoids having to rebuild those methods in our subclass and allows us to swap out superclasses with little code modifications.

Reviewer: ```suggestion
Using `super()` to call already created methods avoids having to rebuild those methods in our subclass and allows us to swap out superclasses with little code modifications.

[four-pillars]: https://www.educative.io/edpresso/what-are-the-four-pillars-of-oops-in-python
```
